dashboard/lib/services/script_seed.rb

```diff
@@ -0,0 +1,320 @@
+  #
+  # Some important implementation details:
+  #
+  # - One challenge is that different models have different ways of uniquely identifying themselves. To abstract these
+  # differences, each model implements a seeding_key method which takes in a SeedContext, and returns a hash of all
+  # the information which composes its "logical unique identifier".
+  #
+  # - We try to avoid doing database queries within loops, within reason. Ideally, the total number of queries made
+  # would only scale with the # of models, not the number of objects. Exceptions can be made if avoiding them is
+  # too complex or not performant, and making the repeated queries achieves decent performance empirically.
+  #
+  # - One way to avoid repeated queries when looking up associations is to load all objects of the associated model
+  # for the Script into the SeedContext, and do lookups against that already-loaded data.
+  #
+  # - If a seeding_key has a corresponding unique index, we can skip looking up objects to update, and just
+  # rely on the "on duplicate key update" feature instead.
+  #
+  # - We try to achieve both simplicity and performance.
+  #
+  # @param [String] json_string - The input JSON to seed from.
+  # @return [Script] the Script created/updated from seeding
+  def self.seed_from_json(json_string)
+    Script.transaction do
+      data = JSON.parse(json_string)
+
+      script_data = data['script']
+      lesson_groups_data = data['lesson_groups']
+      lessons_data = data['lessons']
+      script_levels_data = data['script_levels']
+      levels_script_levels_data = data['levels_script_levels']
+      seed_context = SeedContext.new
+
+      seed_context.script = import_script(script_data)
+      seed_context.lesson_groups = import_lesson_groups(lesson_groups_data, seed_context)
+      seed_context.lessons = import_lessons(lessons_data, seed_context)
+
+      # Because ScriptLevel's seeding_key depends on the keys of its associated Levels, we can
+      # improve performance by loading all of it here into the SeedContext.
+      # TODO: this can be simplified / avoided if we add a new unique identifier field to ScriptLevel.
+      seed_context.script_levels = ScriptLevel.where(script: seed_context.script).includes(:levels)
+      seed_context.levels_script_levels = seed_context.script.levels_script_levels
+      seed_context.script_levels = import_script_levels(script_levels_data, seed_context)
+
+      # levels_script_levels is a join table, which isn't usually explicitly modeled in Rails, but treating it
+      # as just another set of objects to import allows us to handle it with the same pattern as the other models.
+      seed_context.levels = seed_context.script_levels.map(&:levels).flatten
+      import_levels_script_levels(levels_script_levels_data, seed_context)
+      seed_context.script
+    end
+  end
+
+  # Internal methods and classes below
```

Comment:
if you like, it is possible to enforce this for class methods by making them private like this: https://github.com/code-dot-org/code-dot-org/blob/9bc1c06efac1d827bad691e136a3918819326c85/dashboard/app/models/script.rb#L240-L243

Reply:
Hmm... I think I'd go for it if the syntax was a little nicer, but I don't think there's a strong need to enforce that they're private. I mostly wanted to organize the file so it's clear which methods are the ones you should start looking at, and the rest you only need to look at if you're digging in.